lib/signs_ui/alerts/state.ex

```diff
@@ -4,87 +4,147 @@ defmodule SignsUi.Alerts.State do
   end
 
-  @impl GenStage
+  @impl GenServer
   def handle_call(:all, _from, state) do
-    {:reply, Display.format_state(state), [], state}
+    {:reply, Display.format_state(state), state}
   end
 
-  @impl GenStage
-  @spec handle_events([Event.t()], GenStage.from(), t()) :: {:noreply, [], t()}
-  def handle_events(events, _from, state) do
-    # Works in two primary phases. First, we generate fresh t using the
-    # provided events:
-    new_state = update_state(events, state)
-    Logger.info(["alert_state_updated ", inspect(new_state)])
-
-    # Next, we convert our internal model to the specified format:
-    display_state = Display.format_state(new_state)
-    SignsUiWeb.Endpoint.broadcast!("alerts:all", "new_alert_state", display_state)
-    {:noreply, [], new_state}
+  @impl GenServer
+  def handle_info(:update, state) do
+    state =
+      case fetch_alerts() do
+        {:ok, response} ->
+          new_state = Enum.into(parse(response), %{}, &{&1.id, &1})
+
+          SignsUiWeb.Endpoint.broadcast!(
+            "alerts:all",
+            "new_alert_state",
+            Display.format_state(new_state)
+          )
+
+          Logger.info(["alert_state_updated ", inspect(new_state)])
+          new_state
+
+        {:error, error} ->
+          Logger.error([
+            "alerts_fetch_failed, reason=",
+            inspect(error)
+          ])
+
+          state
+      end
+
+    schedule_update(self(), 5_000)
+    {:noreply, state}
   end
 
-  @spec update_state(Event.t() | [Event.t()], t()) :: t()
-  defp update_state(events, state) when is_list(events) do
-    # This reduce combines the effects of a  set of operations into a single new
-    # state.
-    Enum.reduce(events, state, &update_state(&1, &2))
+  defp fetch_alerts do
+    http_client = Application.get_env(:signs_ui, :http_client)
+    url = "#{Application.get_env(:signs_ui, :api_v3_url)}/alerts"
+    headers = [{"x-api-key", Application.get_env(:signs_ui, :api_v3_key)}]
+
+    case http_client.get(
+           url,
+           headers,
+           params: %{
+             "filter[datetime]" => "NOW",
+             "filter[route_type]" => "0,1"
+           }
+         ) do
+      {:ok, %HTTPoison.Response{status_code: 200, body: body}} ->
+        {:ok, body}
+
+      error ->
+        {:error, error}
+    end
   end
 
-  defp update_state(%Event{event: "reset", data: data}, _) do
-    alerts = Events.parse(data)
-    Map.new(alerts, &{&1.id, &1})
+  def parse(payload) when is_binary(payload) do
+    payload |> Jason.decode!() |> Map.get("data", []) |> parse()
   end
 
-  defp update_state(%Event{event: "update", data: data}, state) do
-    alert = Events.parse(data)
-    Map.put(state, alert.id, alert)
+  def parse(payload) when is_list(payload) do
+    Enum.map(payload, &parse/1)
   end
 
-  defp update_state(%Event{event: "add", data: data}, state) do
-    alert = Events.parse(data)
-    Map.put(state, alert.id, alert)
+  def parse(payload) when is_map(payload) do
```

Comment:
nit: Since these `parse` functions operate on different levels of the data structure, it might be slightly clearer to give them distinct names rather than implicitly differentiating by type. (And/or combine some of them)